chain/client/src/chunks_delay_tracker.rs

```diff
@@ -1,192 +1,89 @@
+                    metrics::CHUNK_DELAY
+                        .with_label_values(&[&format!("{}", shard_id)])
+                        .observe(received.saturating_duration_since(*requested).as_secs_f64());
+                }
+            }
         }
     }
 
-    fn update_blocks_ahead_metric(&mut self, head_height: BlockHeight) {
-        metrics::BLOCKS_AHEAD_OF_HEAD.set(self.get_blocks_ahead(head_height) as i64);
-    }
-
-    fn update_metrics(&mut self, head_height: BlockHeight) {
-        self.update_chunks_metric(head_height);
-        self.update_blocks_ahead_metric(head_height);
-    }
-
-    pub fn add_block_timestamp(
-        &mut self,
-        height: BlockHeight,
-        head_height: BlockHeight,
-        timestamp: Instant,
-    ) {
-        self.remove_old_entries(head_height);
-        if height >= head_height {
-            self.heights.entry(height).or_default().block_received.get_or_insert(timestamp);
-        }
-        self.update_metrics(head_height);
+    pub fn add_block_timestamp(&mut self, height: BlockHeight, timestamp: Instant) {
+        self.blocks_in_progress.entry(height).or_default().block_received.get_or_insert(timestamp);
     }
 
     pub fn add_chunk_timestamp(
         &mut self,
         height: BlockHeight,
         shard_id: ShardId,
-        head_height: BlockHeight,
         timestamp: Instant,
     ) {
-        self.remove_old_entries(head_height);
-        if height >= head_height {
-            self.heights
-                .entry(height)
-                .or_default()
-                .chunks_received
-                .entry(shard_id)
-                .or_insert(timestamp);
-        }
-        self.update_metrics(head_height);
+        self.blocks_in_progress
+            .entry(height)
+            .or_default()
+            .chunks_received
+            .entry(shard_id)
+            .or_insert(timestamp);
     }
-}
-#[cfg(test)]
-mod test {
-    use super::*;
-    use std::time::Duration;
-
-    #[test]
-    fn test_blocks_ahead_of_head() {
-        let mut tracker = ChunksDelayTracker::default();
-        let now = Instant::now();
 
-        let mut head_height = 1;
-        tracker.add_block_timestamp(3, head_height, now);
-        assert_eq!(tracker.get_blocks_ahead(head_height), 2);
-        tracker.add_block_timestamp(4, head_height, now);
-        assert_eq!(tracker.get_blocks_ahead(head_height), 3);
-
-        head_height = 3;
-        tracker.add_block_timestamp(4, head_height, now);
-        assert_eq!(tracker.get_blocks_ahead(head_height), 1);
-
-        head_height = 5;
-        tracker.add_block_timestamp(4, head_height, now);
-        assert_eq!(tracker.get_blocks_ahead(head_height), 0);
-
-        head_height = 999;
-        tracker.add_block_timestamp(4, head_height, now);
-        assert_eq!(tracker.get_blocks_ahead(head_height), 0);
+    pub fn processed_block(&mut self, processed_block_height: BlockHeight) {
+        let heights_to_drop: Vec<BlockHeight> =
```

Comment:
I think it's better to just drop the block at the specific height, because blocks before the height may or may not be finished if there are forks. And again, I think it's better to index blocks by hash instead of by height.